File: src/AuthenticationModules.VB/AuthenticationForm2.vb

Friend Class AuthenticationForm2

    Private ReadOnly _Users As User()

    Public ReadOnly Property SelectedUser() As User
        Get
            Return CType(UserInfoComboBox.SelectedItem, User)
        End Get
    End Property

    Sub New()

        InitializeComponent()

        _Users = AuthenticationModuleHelper.ReadUsersFile()

        UserInfoComboBox.DisplayMember = NameOf(User.DisplayName)
        UserInfoComboBox.DataSource = _Users
        UserInfoComboBox.SelectedIndex = -1
    End Sub

    Private Sub UserInfoComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles UserInfoComboBox.SelectedIndexChanged

        OkButton.Enabled = UserInfoComboBox.SelectedIndex >= 0
    End Sub

    Private Sub UserInfoComboBox_MouseDoubleClick(sender As Object, e As MouseEventArgs) Handles UserInfoComboBox.MouseDoubleClick

        OkButton.PerformClick()
    End Sub

    Public Shared Function Authenticate() As User

        Using Dialog = New AuthenticationForm2

            If Dialog.ShowDialog(Form.ActiveForm) = DialogResult.OK Then

                Return Dialog.SelectedUser
            Else

                Return Nothing
            End If
        End Using
    End Function
End Class

File: src/AuthenticationModules.CS/AuthenticationForm2.cs
using System;
using System.Windows.Forms;

namespace AuthenticationModules
{
    partial class AuthenticationForm2
    {
        readonly User[] _users;

        public User SelectedUser => (User)UserInfoComboBox.SelectedItem;

        public AuthenticationForm2()
        {
            InitializeComponent();

            _users = AuthenticationModuleHelper.ReadUsersFile();

            UserInfoComboBox.DisplayMember = nameof(User.DisplayName);
            UserInfoComboBox.DataSource = _users;
            UserInfoComboBox.SelectedIndex = -1;
        }

        void UserInfoComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            OkButton.Enabled = UserInfoComboBox.SelectedIndex >= 0;
        }

        void UserInfoComboBox_MouseDoubleClick(object sender, MouseEventArgs e)
        {
            OkButton.PerformClick();
        }

        public static User Authenticate()
        {
            using (var dialog = new AuthenticationForm2())
            {
                if (dialog.ShowDialog(ActiveForm) == DialogResult.OK)
                {

                    return dialog.SelectedUser;
                }
                else
                {

                    return null;
                }
            }
        }
    }
}